tests/database/test_header_db.py

```diff
@@ -94,18 +96,187 @@ def test_headerdb_persist_disconnected_headers(headerdb, genesis_header):
 
     pseudo_genesis = headers[7]
 
+    assert headerdb.get_header_chain_gaps() == GENESIS_CHAIN_GAPS
     # Persist the checkpoint header with a trusted score
     headerdb.persist_checkpoint_header(pseudo_genesis, score_at_pseudo_genesis)
+    assert headerdb.get_header_chain_gaps() == (((1, 7),), 9)
 
     assert_headers_eq(headerdb.get_canonical_head(), pseudo_genesis)
 
     headers_from_pseudo_genesis = (headers[8], headers[9],)
 
     headerdb.persist_header_chain(headers_from_pseudo_genesis, pseudo_genesis.parent_hash)
+    assert headerdb.get_header_chain_gaps() == (((1, 7),), 11)
     head = headerdb.get_canonical_head()
     assert_headers_eq(head, headers[-1])
     assert headerdb.get_score(head.hash) == expected_score_at_tip
 
+    header_8 = headerdb.get_block_header_by_hash(headers[8].hash)
+    assert_headers_eq(header_8, headers[8])
+
+    with pytest.raises(HeaderNotFound):
+        headerdb.get_block_header_by_hash(headers[2].hash)
+
+
+def test_can_patch_holes(headerdb, genesis_header):
+    headerdb.persist_header(genesis_header)
+    headers = mk_header_chain(genesis_header, length=10)
+
+    assert headerdb.get_header_chain_gaps() == GENESIS_CHAIN_GAPS
+    # Persist the checkpoint header with a trusted score
+    # headers[7] has block number 8 because `headers` doesn't include the genesis
+    pseudo_genesis = headers[7]
+    headerdb.persist_checkpoint_header(pseudo_genesis, 154618822656)
+    assert headerdb.get_header_chain_gaps() == (((1, 7),), 9)
+    assert_headers_eq(headerdb.get_canonical_head(), pseudo_genesis)
+
+    headerdb.persist_header_chain(headers[:7])
+    assert headerdb.get_header_chain_gaps() == ((), 9)
+
+    for number in range(1, 9):
+        # Make sure we can lookup the headers by block number
+        header_by_number = headerdb.get_canonical_block_header_by_number(number)
+        assert header_by_number.block_number == headers[number - 1].block_number
+
+    # Make sure patching the hole does not affect what our current head is
+    assert_headers_eq(headerdb.get_canonical_head(), pseudo_genesis)
+
+
+def test_gap_filling_handles_uncles_correctly(headerdb, genesis_header):
```

Comment:
@carver I found it easier to make two assertions in the same test rather than two individual tests to avoid a big chunk of copy and paste.

Reply:
I'm not familiar with `hypothesis` but I've build something home grown that is very simple yet flexible. Maybe that gets the job done?